File: src/ChpokkWeb/SystemFiles/Templates/ProjectTemplates/VisualBasic/Silverlight/1036/BusinessApplication/Helpers/DataFieldExtensions.vb
Imports System
Imports System.Runtime.CompilerServices
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Data

Public Module DataFieldExtensions
    ''' <summary>
    ''' Remplace le contrôle <see cref="TextBox" /> d'un <see cref="DataField" /> par un autre contrôle et met à jour les liaisons.
    ''' </summary>
    ''' <param name="field"><see cref="DataField"/> dont <see cref="TextBox"/> sera remplacé.</param>
    ''' <param name="newControl">Nouveau contrôle que vous allez définir en tant que <see cref="DataField.Content" />.</param>
    ''' <param name="dataBindingProperty">Propriété du contrôle qui sera utilisée pour la liaison de données.</param>        
    ''' <param name="bindingSetupFunction">
    ''' <c>Action</c> optionnelle que vous pouvez utiliser pour modifier des paramètres sur la liaison qui vient d'être créée avant qu'elle soit appliquée à <paramref name="newControl"/>.
    ''' </param>
    <Extension()> _
    Public Sub ReplaceTextBox(ByVal field As DataField, ByVal newControl As FrameworkElement, ByVal dataBindingProperty As DependencyProperty, _
                              Optional ByVal bindingSetupFunction As Action(Of Binding) = Nothing)
        If field Is Nothing Then
            Throw New ArgumentNullException("field")
        End If

        If newControl Is Nothing Then
            Throw New ArgumentNullException("newControl")
        End If

        ' Construire une nouvelle liaison en en copiant une existante, et en la passant à une fonction donnée par l'utilisateur (le cas échéant)
        Dim newBinding As Binding = field.Content.GetBindingExpression(TextBox.TextProperty).ParentBinding.CreateCopy()

        If bindingSetupFunction IsNot Nothing Then
            bindingSetupFunction(newBinding)
        End If

        ' Remplacer le champ
        newControl.SetBinding(dataBindingProperty, newBinding)
        field.Content = newControl
    End Sub

    ''' <summary>
    ''' Crée un nouvel objet <see cref="Binding"/> en copiant toutes les propriétés d'un autre objet <see cref="Binding"/>.
    ''' </summary>
    ''' <param name="binding"><see cref="Binding"/> à partir duquel les valeurs de propriétés seront copiées</param>
    ''' <returns>Nouvel objet <see cref="Binding"/>.</returns>
    <Extension()> _
    Public Function CreateCopy(ByVal binding As Binding) As binding
        If binding Is Nothing Then
            Throw New ArgumentNullException("binding")
        End If

        Dim newBinding As New Binding()

        newBinding.BindsDirectlyToSource = binding.BindsDirectlyToSource
        newBinding.Converter = binding.Converter
        newBinding.ConverterParameter = binding.ConverterParameter
        newBinding.ConverterCulture = binding.ConverterCulture
        newBinding.Mode = binding.Mode
        newBinding.NotifyOnValidationError = binding.NotifyOnValidationError
        newBinding.Path = binding.Path
        newBinding.UpdateSourceTrigger = binding.UpdateSourceTrigger
        newBinding.ValidatesOnExceptions = binding.ValidatesOnExceptions

        If binding.ElementName IsNot Nothing Then
            newBinding.ElementName = binding.ElementName
        ElseIf binding.RelativeSource IsNot Nothing Then
            newBinding.RelativeSource = binding.RelativeSource
        Else
            newBinding.Source = binding.Source
        End If

        Return newBinding
    End Function
End Module

File: src/ChpokkWeb/SystemFiles/Templates/ProjectTemplates/CSharp/Silverlight/1049/BusinessApplication/Helpers/DataFieldExtensions.cs
namespace $safeprojectname$
{
    using System;
    using System.Windows;
    using System.Windows.Controls;
    using System.Windows.Data;

    /// <summary>
    ///     Содержит методы расширения для выполнения операций над объектом <see cref="DataField"/>.
    /// </summary>
    public static class DataFieldExtensions
    {
        /// <summary>
        /// Заменяет элемент управления <see cref="TextBox" /> для <see cref="DataField" /> другим элементом управления и обновляет привязку.
        /// </summary>
        /// <param name="field">Объект <see cref="DataField"/>, элемент управления <see cref="TextBox"/> которого нужно заменить.</param>
        /// <param name="newControl">Новый элемент управления, который должен быть назначен в качестве <see cref="DataField.Content" />.</param>
        /// <param name="dataBindingProperty">Свойство элемента управления, которое служит для привязки данных.</param>        
        public static void ReplaceTextBox(this DataField field, FrameworkElement newControl, DependencyProperty dataBindingProperty)
        {
            field.ReplaceTextBox(newControl, dataBindingProperty, null);
        }

        /// <summary>
        /// Заменяет элемент управления <see cref="TextBox" /> для <see cref="DataField" /> другим элементом управления и обновляет привязку.
        /// </summary>
        /// <param name="field">Объект <see cref="DataField"/>, элемент управления <see cref="TextBox"/> которого нужно заменить.</param>
        /// <param name="newControl">Новый элемент управления, который должен быть назначен в качестве <see cref="DataField.Content" />.</param>
        /// <param name="dataBindingProperty">Свойство элемента управления, которое служит для привязки данных.</param>        
        /// <param name="bindingSetupFunction">
        ///  Необязательный параметр <see cref="Action"/>, который может быть использован для изменения параметров во вновь созданной привязке, прежде чем она будет применена к <paramref name="newControl"/>
        /// </param>
        public static void ReplaceTextBox(this DataField field, FrameworkElement newControl, DependencyProperty dataBindingProperty, Action<Binding> bindingSetupFunction)
        {
            if (field == null)
            {
                throw new ArgumentNullException("field");
            }

            if (newControl == null)
            {
                throw new ArgumentNullException("newControl");
            }

            // Создает новую привязку путем копирования существующей и передачи ее функции bindingSetupFunction, в которой в вызывающем коде можно произвести любые необходимые изменения.
            Binding newBinding = field.Content.GetBindingExpression(TextBox.TextProperty).ParentBinding.CreateCopy();

            if (bindingSetupFunction != null)
            {
                bindingSetupFunction(newBinding);
            }

            // Замена поля
            newControl.SetBinding(dataBindingProperty, newBinding);
            field.Content = newControl;
        }

        /// <summary>
        /// Создает новый объект <see cref="Binding"/>, копируя все свойства из другого объекта <see cref="Binding"/>.
        /// </summary>
        /// <param name="binding"><see cref="Binding"/>, из которого будут скопированы значения свойств</param>
        /// <returns>Новый объект <see cref="Binding"/>.</returns>
        private static Binding CreateCopy(this Binding binding)
        {
            if (binding == null)
            {
                throw new ArgumentNullException("binding");
            }

            Binding newBinding = new Binding()
            {
                BindsDirectlyToSource = binding.BindsDirectlyToSource,
                Converter = binding.Converter,
                ConverterParameter = binding.ConverterParameter,
                ConverterCulture = binding.ConverterCulture,
                Mode = binding.Mode,
                NotifyOnValidationError = binding.NotifyOnValidationError,
                Path = binding.Path,
                UpdateSourceTrigger = binding.UpdateSourceTrigger,
                ValidatesOnExceptions = binding.ValidatesOnExceptions
            };

            if (binding.ElementName != null)
            {
                newBinding.ElementName = binding.ElementName;
            }
            else if (binding.RelativeSource != null)
            {
                newBinding.RelativeSource = binding.RelativeSource;
            }
            else
            {
                newBinding.Source = binding.Source;
            }

            return newBinding;
        }
    }
}
